pkg/controllers/monitoring-stack/components.go

```diff
@@ -0,0 +1,403 @@
+package monitoringstack
+
+import (
+	"fmt"
+	stack "rhobs/monitoring-stack-operator/pkg/apis/v1alpha1"
+	grafana_operator "rhobs/monitoring-stack-operator/pkg/controllers/grafana-operator"
+
+	grafanav1alpha1 "github.com/grafana-operator/grafana-operator/v4/api/integreatly/v1alpha1"
+
+	monv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const AdditionalScrapeConfigsSelfScrapeKey = "self-scrape-config"
+
+type emptyObjectFunc func() client.Object
+
+type patchObjectFunc func(object client.Object) (client.Object, error)
+
+type objectPatcher struct {
+	empty emptyObjectFunc
+	patch patchObjectFunc
+}
+
+type ErrInvalidObjectType struct {
+	expectedType string
+}
+
+func NewErrInvalidObjectType(expectedType string) error {
+	return ErrInvalidObjectType{expectedType: expectedType}
+}
+
+func (e ErrInvalidObjectType) Error() string {
+	return fmt.Sprintf("object type is not %s", e.expectedType)
+}
+
+func (r *reconciler) stackComponentPatchers(ms *stack.MonitoringStack, instanceSelectorKey string, instanceSelectorValue string) ([]objectPatcher, error) {
+	rbacResourceName := ms.Name + "-prometheus"
+	rbacVerbs := []string{"get", "list", "watch"}
+	additionalScrapeConfigsSecretName := ms.Name + "-prometheus-additional-scrape-configs"
+
+	return []objectPatcher{
+		{
+			empty: func() client.Object {
+				sa := newServiceAccount(ms)
+				return &corev1.ServiceAccount{
+					TypeMeta:   sa.TypeMeta,
+					ObjectMeta: sa.ObjectMeta,
+				}
+			},
+			patch: func(object client.Object) (client.Object, error) {
+				return newServiceAccount(ms), nil
+			},
+		},
+		{
+			empty: func() client.Object {
+				role := newRole(ms, rbacResourceName, rbacVerbs)
+				return &rbacv1.Role{
+					TypeMeta:   role.TypeMeta,
+					ObjectMeta: role.ObjectMeta,
+				}
+			},
+			patch: func(object client.Object) (client.Object, error) {
+				role := newRole(ms, rbacResourceName, rbacVerbs)
+
+				if object == nil {
+					return role, nil
+				}
+
+				existing, ok := object.(*rbacv1.Role)
+				if !ok {
+					return nil, NewErrInvalidObjectType("Role.rbac.authorization.k8s.io/v1")
+				}
+
+				existing.Rules = role.Rules
+				return existing, nil
+			},
+		},
+		{
+			empty: func() client.Object {
+				rb := newRoleBinding(ms, rbacResourceName)
+				return &rbacv1.RoleBinding{
+					TypeMeta:   rb.TypeMeta,
+					ObjectMeta: rb.ObjectMeta,
+				}
+			},
+			patch: func(object client.Object) (client.Object, error) {
+				roleBinding := newRoleBinding(ms, rbacResourceName)
+
+				if object == nil {
+					return roleBinding, nil
+				}
+
+				existing, ok := object.(*rbacv1.RoleBinding)
+				if !ok {
+					return nil, NewErrInvalidObjectType("RoleBinding.rbac.authorization.k8s.io/v1")
```

Comment:
this can be simplified to avoid typo by passing the `empty` obj ... e.g
```
func NewErrInvalidObjectType(obj client.Object) error {
	gvk := obj.GetObjectKind().GroupVersionKind()
	gvkString := gvk.Group + "/" + gvk.Version + " " + gvk.Kind

	return ErrInvalidObjectType{expectedType: gvkString}
}
...

return nil, NewErrInvalidObjectType(roleBinding)

```
while at it, how about adding information about what object was actually passed and not just what was expected?